File: dnkdialer-android/app/src/test/java/com/dnkdialer/companion/E2eSasBlockingContractTest.kt
package com.dnkdialer.companion

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * SAS-MODE0 — the phone half of the cross-surface contract
 * (RESUME-PROTOCOL v3.0 RULE 30).
 *
 * ## One file, two implementations
 *
 * `tests/e2e-sas-blocking-vectors.json` at the repo root is the SINGLE source
 * of truth for the SPEC §13.2 cells. The web lane asserts it in
 * `tests/e2e-sas-blocking-contract.test.mjs` (decideAccept -> the view useE2e
 * publishes -> sasIsBlocking); this file asserts the SAME rows against
 * [E2eSettings.effectiveMode] / [E2eSettings.isSealed] /
 * [E2eSettings.requiresSas] and [E2eStatusCopy.stateOf] /
 * [E2eStatusCopy.statusLine]. Same pattern as [E2eKdfVectorsTest]: a shared
 * vector file is the only thing that can catch the two sides quietly
 * disagreeing, because each side's own tests agree with it by construction.
 *
 * ## The row this exists for
 *
 * `row-4-M1`: phone encrypted mode OFF, computer encrypted mode OFF, a usable
 * e2e block on both sides. The pair SEALS at modeByte 0x00, the EFFECTIVE mode
 * stays OFF, SAS digits are computed (frozen transcript + coverage) and NOBODY
 * is asked to confirm them. Both surfaces must read "Encrypted, unverified"
 * and NEITHER may block on a code.
 *
 * Live acceptance of web 3e466fd (2026-09-23) found the browser opening the
 * blocking SAS modal on exactly this pair while the phone showed no code. That
 * was a web reader bug (`sasIsBlocking` keyed on the SEALING flag) and it is
 * fixed in the web lane. The decision table on THIS side was already right,
 * and this test pins it so it stays right.
 *
 * ## What this test deliberately does NOT cover
 *
 * The phone's live status line read "Connected · Not encrypted" on row 4. That
 * is NOT this table: it is the WIRING above it. `E2eTofuContract.ACTION_E2E_STATE`
 * — the broadcast MainActivity listens for at MainActivity.kt:396 to move
 * `e2eState` off its PLAINTEXT default — has no sender anywhere in
 * `app/src/main`; the only thing that fires it is an instrumented test. So the
 * line can only leave PLAINTEXT via the optimistic set on a CONFIRMED SAS
 * (MainActivity.kt:2415), which by definition never happens on row 4.
 *
 * That is ticket T-PHONE-STATUS-MODE0 and belongs to the vc63 android lane.
 * This lane changes no Kotlin product code (Ken's ruling), so the finding is
 * recorded here and in the résumé rather than fixed.
 *
 * Run: `gradlew.bat testDebugUnitTest --tests '*SasBlockingContract*'`
 */
class E2eSasBlockingContractTest {

    /** Unit-test cwd is `dnkdialer-android/app`, so `../..` is the repo root. */
    private val file = File("../../tests/e2e-sas-blocking-vectors.json")

    private fun rows(): List<JsonObject> {
        assertTrue(
            "the shared vector file is missing at " + file.absolutePath +
                " — the web lane and this one must read the SAME file",
            file.exists()
        )
        val root = JsonParser.parseString(file.readText()).asJsonObject
        assertEquals("vector file version", 1, root.get("version").asInt)
        val rows = root.getAsJsonArray("rows").map { it.asJsonObject }
        // A vectors test whose file lost its rows passes vacuously. It must not.
        assertEquals("the §13.2 matrix is 2x2x2 — every cell is present", 8, rows.size)
        return rows
    }

    /** The peer the PHONE sees: the computer, through its advertised block. */
    private fun peerOf(row: JsonObject): E2eSettings.PeerAdvertisement =
        if (!row.get("blockPresent").asBoolean) E2eSettings.PeerAdvertisement.ABSENT
        else if (row.get("computerEncryptedMode").asString == "on") E2eSettings.PeerAdvertisement.ON
        else E2eSettings.PeerAdvertisement.OFF

    @Test
    fun every_matrix_cell_matches_the_shared_vectors() {
        var sealedRows = 0
        for (row in rows()) {
            val id = row.get("id").asString
            val phone = row.getAsJsonObject("phone")
            val localEnabled = row.get("phoneEncryptedMode").asString == "on"

            val mode = E2eSettings.effectiveMode(localEnabled, peerOf(row))
            assertEquals("$id: effectiveMode", phone.get("effectiveMode").asString, mode.name)
            assertEquals("$id: isSealed", phone.get("sealed").asBoolean, E2eSettings.isSealed(mode))
            assertEquals(
                "$id: requiresSas", phone.get("requiresSas").asBoolean,
                E2eSettings.requiresSas(mode)
            )

            // The SAS is blocking on this side exactly when the effective mode
            // requires it AND the pair actually sealed — the phone never shows
            // a code for a pair it is not encrypting.
            val blocking = E2eSettings.isSealed(mode) && E2eSettings.requiresSas(mode)
            assertEquals("$id: sasBlocking", phone.get("sasBlocking").asBoolean, blocking)

            if (!E2eSettings.isSealed(mode)) {
                // ABORT has no status line at all; PLAINTEXT has the unencrypted one.
                if (mode == E2eSettings.EffectiveMode.ABORT) {
                    assertTrue("$id: an aborted pair has no status line", phone.get("state").isJsonNull)
                } else {
                    assertEquals(
                        "$id: statusLine", R.string.status_connected_unencrypted,
                        E2eStatusCopy.statusLine(E2eStatusCopy.stateOf(false, false))
                    )
                }
                continue
            }

            sealedRows++
            val state = E2eStatusCopy.stateOf(
                encrypted = true,
                verified = E2eSettings.requiresSas(mode),
            )
            assertEquals("$id: E2eStatusCopy state", phone.get("state").asString, state.name)
            val expectedRes = when (phone.get("statusLineKey").asString) {
                "status_connected_encrypted" -> R.string.status_connected_encrypted
                "status_connected_encrypted_unverified" -> R.string.status_connected_encrypted_unverified
                "status_connected_unencrypted" -> R.string.status_connected_unencrypted
                else -> throw AssertionError("$id: unknown statusLineKey")
            }
            assertEquals("$id: statusLine", expectedRes, E2eStatusCopy.statusLine(state))
            // Never colour-only, and never silently the same string for two states.
            assertTrue(
                "$id: verified and unverified must not collapse to one string",
                R.string.status_connected_encrypted != R.string.status_connected_encrypted_unverified
            )
        }
        assertEquals("four of the eight cells seal", 4, sealedRows)
    }

    /**
     * Row 4 again, spelled out by hand rather than driven from the file.
     *
     * A loop over vectors proves the implementation matches the file. It does
     * not prove the FILE still says the thing the row exists to say — deleting
     * row 4 would leave the loop green. So the claim is also written here, in
     * Kotlin, where a vector-file edit cannot reach it.
     */
    @Test
    fun row_four_seals_with_the_sas_not_blocking_on_this_side() {
        val mode = E2eSettings.effectiveMode(
            localEnabled = false,
            peer = E2eSettings.PeerAdvertisement.OFF,
        )
        assertEquals(E2eSettings.EffectiveMode.ENCRYPTED_UNVERIFIED, mode)
        assertTrue("a 0/0 pair with a capable peer SEALS", E2eSettings.isSealed(mode))
        assertFalse("and nobody is asked to confirm a code", E2eSettings.requiresSas(mode))
        assertEquals(
            "the phone must say Encrypted, unverified — not Not encrypted",
            R.string.status_connected_encrypted_unverified,
            E2eStatusCopy.statusLine(E2eStatusCopy.stateOf(encrypted = true, verified = false)),
        )
        // The control: if either side had asked, this same cell WOULD block.
        // If this stops holding, the assertion above has stopped measuring.
        val asked = E2eSettings.effectiveMode(false, E2eSettings.PeerAdvertisement.ON)
        assertTrue("row 8 control: a peer that asked makes the SAS blocking",
            E2eSettings.requiresSas(asked))
    }

    /**
     * T-PHONE-STATUS-MODE0, pinned as an OBSERVATION so the android lane can
     * find it and so nobody re-derives it from a live run.
     *
     * Asserted as "no production sender" rather than fixed: this lane ships no
     * Kotlin product code. When the android lane wires the broadcast, this test
     * goes red and is deleted in the same commit — which is the point.
     */
    @Test
    fun observation_nothing_in_production_broadcasts_the_e2e_state() {
        val main = File("src/main/java/com/dnkdialer/companion")
        assertTrue("main sources not found at " + main.absolutePath, main.isDirectory)
        val senders = main.walkTopDown()
            .filter { it.isFile && it.extension == "kt" }
            // The only shape that can SEND it: an Intent built on the action.
            // Matching the bare constant would count MainActivity's RECEIVER
            // (MainActivity.kt:396 / the IntentFilter at :2224) as a sender.
            .filter { f -> f.readText().contains("Intent(E2eTofuContract.ACTION_E2E_STATE)") }
            .map { it.name }
            .toList()
        assertEquals(
            "T-PHONE-STATUS-MODE0: a production sender for ACTION_E2E_STATE appeared (" +
                senders + "). The status line can now leave PLAINTEXT on its own — " +
                "delete this observation test and assert the real wiring instead.",
            emptyList<String>(), senders
        )
    }
}

File: tests/e2e-sas-blocking-contract.test.mjs
/**
 * tests/e2e-sas-blocking-contract.test.mjs — SAS-MODE0, RESUME-PROTOCOL RULE 30.
 *
 * ONE vector file, `tests/e2e-sas-blocking-vectors.json`, consumed by TWO
 * implementations:
 *
 *   - this file — the REAL web path: `decideAccept` (hooks/phoneE2e.ts) over a
 *     real `e2e` block, reduced to the view shape `hooks/useE2e.ts` actually
 *     publishes, fed to `sasIsBlocking` and `encryptionIndicator`
 *     (lib/encryptedModeCopy.ts);
 *   - dnkdialer-android/app/src/test/java/com/dnkdialer/companion/
 *     E2eSasBlockingContractTest.kt — `E2eSettings.effectiveMode/isSealed/
 *     requiresSas` and `E2eStatusCopy.stateOf/statusLine` over the SAME rows.
 *
 * ## Why the view shape is reconstructed here and not imported
 *
 * `useE2e` is a React hook; the fact under test is the ONE line it writes:
 *
 *     setView((v) => ({ ...v, mode: 'on', effective: decision.effective, ... }))
 *
 * `mode` is hard-coded 'on' for every sealed pair — it is the SEALING flag —
 * while `effective` carries the decision. That is the exact conflation that
 * produced the SAS-MODE0 defect, so this suite does not take it on trust: it
 * ASSERTS the literal `mode: 'on'` and `effective: decision.effective` out of
 * the hook's source (comments stripped) before it uses the shape. A contract
 * test built on a view shape nobody checked against the hook proves the vector
 * parser, not the product.
 *
 * ## The row this exists for
 *
 * row-4-M1: phone OFF, computer OFF, usable block. Sealed, effective OFF,
 * digits computed, NOBODY asked to verify. Live acceptance 3e466fd found the
 * blocking modal over the whole panel on exactly this pair because
 * `sasIsBlocking` read `view.mode`.
 *
 * Run: node tests/e2e-sas-blocking-contract.test.mjs
 */

import { createRequire } from 'node:module';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';

import { readAcceptBlock, decideAccept } from '../hooks/phoneE2e.ts';
import { sasIsBlocking, encryptionIndicator } from '../lib/encryptedModeCopy.ts';

const require = createRequire(import.meta.url);
const VECTORS = require('./e2e-sas-blocking-vectors.json');
const ROOT = path.resolve(import.meta.dirname, '..');

let passed = 0;
let failed = 0;
function check(name, ok, detail = '') {
  if (ok) { passed++; return true; }
  failed++;
  console.error(`  FAIL  ${name}${detail ? ` — ${detail}` : ''}`);
  return false;
}
const eq = (name, got, want) =>
  check(name, got === want, `got ${JSON.stringify(got)} want ${JSON.stringify(want)}`);

/** CRLF-safe (.gitattributes pins lib/hooks to LF; a checkout may not be). */
function stripComments(src) {
  return src
    .replace(/\r\n/g, '\n')
    .replace(/\/\*[\s\S]*?\*\//g, '')
    .replace(/^[ \t]*\/\/.*$/gm, '')
    .replace(/[ \t]\/\/.*$/gm, '');
}

// ── 0. the shape this suite reconstructs is the shape the hook publishes ───
{
  const hook = stripComments(readFileSync(path.join(ROOT, 'hooks', 'useE2e.ts'), 'utf8'));
  check("useE2e still publishes the SEALING flag as a hard-coded mode:'on'",
    /setView\(\(v\) => \(\{[\s\S]{0,400}?mode: 'on',/.test(hook));
  check('useE2e still publishes effective from the decision, not from mode',
    /effective: decision\.effective,/.test(hook));
  check('sasPendingRef (the frame chokepoint) keys on the EFFECTIVE mode',
    /sasPendingRef\.current = decision\.effective === 'on'/.test(hook));

  const copy = stripComments(readFileSync(path.join(ROOT, 'lib', 'encryptedModeCopy.ts'), 'utf8'));
  // The BODY only. `fn.indexOf('\n}')` would stop at the `}): boolean {` that
  // closes the PARAMETER type literal — also at column 0 — and leave an empty
  // fragment that passes the negative assertion for free.
  const fn = copy.slice(copy.indexOf('export function sasIsBlocking'));
  const open = fn.indexOf('): boolean {');
  const body = fn.slice(open, fn.indexOf('\n}', open) + 2);
  check('the sasIsBlocking body actually sliced (not an empty fragment)',
    body.includes('return ') && body.length > 60, JSON.stringify(body));
  check('sasIsBlocking gates on view.effective', /view\.effective !== 'on'/.test(body));
  check('sasIsBlocking never reads view.mode (the sealing flag)',
    !/view\.mode/.test(body), body);
}

// ── real key material, so readAcceptBlock accepts the block ────────────────
const b64u = (b) => Buffer.from(b).toString('base64')
  .replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
function p256() {
  const ec = crypto.createECDH('prime256v1');
  ec.generateKeys();
  return { b64: b64u(ec.getPublicKey()) };
}
const PHONE = p256(); const WEB = p256(); const SW = p256(); const EPK = p256();
const WEB_ID = 'web-dev-0000000000000001';
const SW_ID = 'ext-dev-0000000000000002';
const wrapFor = (deviceId) => ({ deviceId, wrap: b64u(crypto.randomBytes(48)) });

/**
 * The phone's `e2e` block for a row. `mode` is the PHONE's own setting as the
 * byte — that is what §13.3 puts on the wire and what `effectiveMode` ORs with
 * the local setting. `blockPresent:false` -> null (the pre-feature peer).
 */
function blockFor(row) {
  if (!row.blockPresent) return null;
  return readAcceptBlock({
    v: 1,
    mode: row.phoneEncryptedMode === 'on' ? 1 : 0,
    kid: `kid-${row.id}`,
    epk: EPK.b64,
    recipKeys: [PHONE.b64, WEB.b64, SW.b64],
    wraps: [wrapFor(WEB_ID), wrapFor(SW_ID)],
  });
}

/** The view `useE2e` publishes for a decision — the shape asserted in step 0. */
function viewFor(decision, digits) {
  if (decision.action === 'abort') {
    return {
      mode: decision.mode, effective: decision.effective, state: 'error',
      sas: { digits: null, confirmed: false },
    };
  }
  if (decision.state === 'unencrypted') {
    return { mode: 'off', effective: 'off', state: 'unencrypted', sas: { digits: null, confirmed: false } };
  }
  // Sealed. Digits exist for EVERY sealed pair (frozen transcript + coverage),
  // including M1 — the defect was never that they are absent.
  return {
    mode: 'on', effective: decision.effective, state: decision.state,
    sas: { digits, confirmed: false },
  };
}

console.log(`SAS-blocking contract — ${VECTORS.rows.length} rows from tests/e2e-sas-blocking-vectors.json`);
for (const row of VECTORS.rows) {
  const d = decideAccept({
    localMode: row.computerEncryptedMode,
    block: blockFor(row),
    ourDeviceId: WEB_ID,
    phoneRowPublicKey: PHONE.b64,
    latched: false,
  });
  eq(`${row.id}: action`, d.action, row.web.action);
  eq(`${row.id}: effective`, d.effective, row.web.effective);
  if (row.web.error) eq(`${row.id}: error`, d.error, row.web.error);

  const view = viewFor(d, VECTORS.digitsForSealedPairs);
  eq(`${row.id}: view.mode (SEALING flag)`, view.mode, row.web.mode);
  eq(`${row.id}: view.state`, view.state, row.web.state);
  eq(`${row.id}: view.sas.digits`, view.sas.digits, row.web.digits);
  eq(`${row.id}: sealed`,
    view.state === 'encrypted-verified' || view.state === 'encrypted-unverified',
    row.web.sealed);
  eq(`${row.id}: sasIsBlocking`, sasIsBlocking(view), row.web.sasBlocking);

  // The banner/header the user reads for the same cell, so a green blocking
  // verdict cannot coexist with copy that contradicts it.
  const ind = encryptionIndicator({ state: view.state, error: d.error, peer: { supports: true } });
  check(`${row.id}: the indicator names a state`,
    typeof ind.label === 'string' && ind.label.length > 0);
  if (row.web.state === 'encrypted-unverified') {
    check(`${row.id}: unverified copy says unverified, never plain "Encrypted"`,
      /unverified|nobody confirmed/i.test(`${ind.label} ${ind.detail}`),
      `${ind.label} / ${ind.detail}`);
  }
}

// ── the defect, stated once more as its own named assertion ────────────────
{
  const row = VECTORS.rows.find((r) => r.id === 'row-4-M1-both-off-block-present');
  check('row 4 exists in the vector file', Boolean(row));
  const d = decideAccept({
    localMode: 'off', block: blockFor(row), ourDeviceId: WEB_ID,
    phoneRowPublicKey: PHONE.b64, latched: false,
  });
  const view = viewFor(d, '76386');
  eq('SAS-MODE0: a 0/0 pair with a usable block SEALS', view.mode, 'on');
  eq('SAS-MODE0: its effective mode stays OFF', view.effective, 'off');
  eq('SAS-MODE0: its digits exist', view.sas.digits, '76386');
  eq('SAS-MODE0: and the modal does NOT open', sasIsBlocking(view), false);
  // The control that makes this suite capable of going red: re-key the gate on
  // the SEALING flag and this very pair blocks again. If this line ever stops
  // returning true, the assertion above has stopped measuring anything.
  eq('SAS-MODE0: keying on the sealing flag WOULD have blocked (control)',
    sasIsBlocking({ ...view, effective: view.mode }), true);
}

console.log(`\n${failed === 0 ? 'PASS' : 'FAIL'}  ${passed} passed, ${failed} failed`);
process.exit(failed === 0 ? 0 : 1);
